src/mscorlib/src/System/Text/ASCIIEncoding.cs

```diff
@@ -763,13 +763,46 @@ public override bool IsSingleByte
 
         public override Decoder GetDecoder()
         {
-            return new DecoderNLS(this);
+            return new ASCIIDecoder(this);
         }
 
-
         public override Encoder GetEncoder()
         {
-            return new EncoderNLS(this);
+            return new ASCIIEncoder(this);
+        }
```

Comment:
When if type information needs to propagate from callee to caller (eg a method returning an exact or sealed subtype of its declared return type) then it can lead to devirtualization but not inlining.

I would very much like to fix this, but it is lower priority than fixing cases where the jit loses types or doesn't get the best type information available.